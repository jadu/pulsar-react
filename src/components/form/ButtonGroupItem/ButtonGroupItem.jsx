import React from 'react';
import classnames from 'classnames';
import FormLabel from '../FormLabel';

export default class ButtonGroupItem extends React.Component {

  static defaultProps = {
    className: 'form__control',
    type: 'radio',
    required: false
  };

  render() {
    let {
      ariaDescribedby,
      children,
      className,
      id,
      idGuid,
      type,
      ...props
    } = this.props;

    let inputClassName = classnames(className, type);

    return (
      <>
        <input 
          className={inputClassName}
          id={id ? id : idGuid} 
          type={type}
          aria-describedby={ariaDescribedby}
          {...props} 
        />
        <FormLabel
          htmlFor={id ? id : idGuid}
        >
          {children}
        </FormLabel>
      </>
    );
  }
}