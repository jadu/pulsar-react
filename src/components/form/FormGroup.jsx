import React from 'react';
import classnames from 'classnames';
import shortid from 'shortid';

import Fieldset from './Fieldset';
import FormLabel from './FormLabel';
import ErrorBlock from './ErrorBlock';
import HelpBlock from './HelpBlock';

export default class FormGroup extends React.Component {

  static defaultProps = {
    className: 'form__group',
    required: false
  };

  componentWillMount() {
    this.guid = shortid.generate();
  }

  render() {
    if (!this.props.children) {
      return null;
    }

    let {
      appendText,
      buttonGroup,
      changed,
      checkbox,
      children,
      choiceGroup,
      className,
      error,
      flushLabel,
      helpText,
      indented,
      inlineCheckbox,
      inlineRadioButton,
      labelText,
      prependText,
      required,
      radio,
      success,
      toggle,
      topLabel,
      warning,
      width,
      ...props} = this.props;

    // Convert variants to their required classes
    let variantClasses = classnames(className, {
      [`form__control-col--${width}`]: width,
      'form__button-group': buttonGroup,
      'form-choice': choiceGroup,
      'form__group--indent': indented,
      'form__group--toggle': toggle,
      'form__group--top': topLabel,
      'form__group--flush': flushLabel,
      'form-checkbox': checkbox,
      'form-checkbox-inline': inlineCheckbox,
      'form-radio-inline': inlineRadioButton,
      'form-radio': radio,
      'has-changed': changed,
      'has-success': success,
      'has-warning': warning,
      'has-error': error
    });

    // GUID to use if no explicit ID has been set
    let idGuid = 'id-guid-' + this.guid;
    
    // GUID to use to link the input with any help text 
    let helpGuid = 'help-guid-' + this.guid;
    
    // GUID to use to link the input with any errors 
    let errorGuid = 'error-guid-' + this.guid;

    // Build list of IDs for the input based on the presence of help/error text 
    let ariaDescribedby = classnames({
      [`${errorGuid}`]: error,
      [`${helpGuid}`]: this.props.helpText
    });

    // If help/errors are present pass the GUIDs to the child inputs
    let childrenWithGuids = React.cloneElement(children, {
      ariaDescribedby: ariaDescribedby ? ariaDescribedby : null,
      idGuid: idGuid
    });

    // Build list of classNames to be used on the controls block
    let controlsClassName = classnames('controls', {
      'btn__group': buttonGroup
    });

    // Form controls
    let inputBlock = (
      <>{childrenWithGuids}</>
    );

    // Wrap with input group if appended/prepended text is present
    if (appendText || prependText) {
      inputBlock = (
        <div className="input-group">
          {prependText && 
            <span className="input-group-addon">{prependText}</span>
          }
          {childrenWithGuids}
          {appendText && 
            <span className="input-group-addon">{appendText}</span>
          }
        </div>
      );
    }

    // Wrap input block alongside errors and help
    let controlsBlock = (
      <div className={controlsClassName}>
        {inputBlock}
        <ErrorBlock errorGuid={errorGuid}>
          {error}
        </ErrorBlock> 
        <HelpBlock helpGuid={helpGuid}>
          {helpText}
        </HelpBlock>
      </div>
    );

    // Standard group markup strategy for most components
    let groupBlock = (
      <div className={variantClasses}>
        <FormLabel 
          required={required} 
          idGuid={idGuid}
        >
          {labelText}
        </FormLabel>
        {controlsBlock}
      </div>
    );

    // Switch markup strategies for certain non-standard components
    if (choiceGroup) {
      groupBlock = (
        <Fieldset 
          className={variantClasses}
          legendText={labelText}
          legendClassName="control__label"
        >
          {controlsBlock}
        </Fieldset>
      );
    }
    else if (toggle) {
      groupBlock = (
        <div className={variantClasses}>
          <FormLabel 
            className="toggle-switch-wrapper-label" 
            required={required} 
            idGuid={idGuid}
          >
            <span className="control__label">
              {labelText}
            </span>
            {controlsBlock}
          </FormLabel>
        </div>
      );
    } 
    else if (inlineCheckbox || inlineRadioButton) {
      groupBlock = (
        <div className={variantClasses}>
          {controlsBlock}
          {labelText}
        </div>
      );
    }

    return (
      <>{groupBlock}</>
    );
  }
}