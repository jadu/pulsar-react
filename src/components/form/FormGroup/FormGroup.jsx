import React from 'react';
import classnames from 'classnames';
import shortid from 'shortid';

import Fieldset from '../Fieldset/Fieldset';
import FormLabel from '../FormLabel/FormLabel';
import ErrorBlock from '../ErrorBlock/ErrorBlock';
import HelpBlock from '../HelpBlock/HelpBlock';

export default class FormGroup extends React.Component {

  static defaultProps = {
    className: 'form__group',
    required: false
  };

  render() {
    if (!this.props.children) {
      return null;
    }

    let {
      appendText,
      buttonGroup,
      changed,
      checkbox,
      children,
      choiceGroup,
      className,
      controls,
      error,
      flushLabel,
      helpText,
      hideLabel,
      indented,
      inline,
      inlineCheckbox,
      inlineRadioButton,
      labelText,
      prependText,
      required,
      radioButton,
      rangeInput,
      success,
      toggle,
      topLabel,
      warning,
      width,
      ...props} = this.props;

    // Convert variants to their required classes
    let variantClasses = classnames(className, {
      [`form__control-col--${width}`]: width,
      'form__button-group': buttonGroup,
      'form-choice': choiceGroup,
      'form__group--compound': inline,
      'form__group--indent': indented,
      'form__group--toggle': toggle,
      'form__group--top': topLabel,
      'form__group--flush': flushLabel,
      'form-checkbox': checkbox,
      'form-checkbox-inline': inlineCheckbox,
      'form-radio-inline': inlineRadioButton,
      'form-radio': radioButton,
      'form-range': rangeInput,
      'has-changed': changed,
      'has-success': success,
      'has-warning': warning,
      'has-error': error
    });

    let guid = this.props.guid || shortid.generate();

    // GUID to use if no explicit ID has been set
    let idGuid = 'id-guid-' + guid;
    
    // GUID to use to link the input with any help text 
    let helpGuid = 'help-guid-' + guid;
    
    // GUID to use to link the input with any errors 
    let errorGuid = 'error-guid-' + guid;

    // Build list of IDs for the input based on the presence of help/error text 
    let ariaDescribedby = classnames({
      [`${errorGuid}`]: error,
      [`${helpGuid}`]: this.props.helpText
    });

    // If help/errors are present pass the GUIDs to the child inputs
    let childrenWithGuids = React.Children.map(
      children,
      (child, i) => {
        if (['div', 'p'].indexOf(child.type) !== -1) {
          return child;
        }

        return React.cloneElement(child, {
          ariaDescribedby: ariaDescribedby ? ariaDescribedby : null,
          id: idGuid
        });
      }
    );

    // Build list of classNames to be used on the controls block
    let controlsClassName = classnames('controls', {
      'btn__group': buttonGroup
    });

    // Form controls
    let inputBlock = (
      <React.Fragment>{childrenWithGuids}</React.Fragment>
    );

    // Wrap with input group if appended/prepended text is present
    if (appendText || prependText) {
      inputBlock = (
        <div className="input-group">
          {prependText && 
            <span className="input-group-addon">{prependText}</span>
          }
          {childrenWithGuids}
          {appendText && 
            <span className="input-group-addon">{appendText}</span>
          }
        </div>
      );
    }

    // Wrap input block alongside errors and help
    let controlsBlock = (
      <div className={controlsClassName}>
        {inputBlock}
        <ErrorBlock errorGuid={errorGuid}>
          {error}
        </ErrorBlock> 
        <HelpBlock helpGuid={helpGuid}>
          {helpText}
        </HelpBlock>
      </div>
    );

    // Standard group markup strategy for most components
    let groupBlock = (
      <div className={variantClasses}>
        {labelText &&
        <FormLabel 
          required={required} 
          idGuid={idGuid}
          tag={controls ? 'span' : 'label'}
          hideLabel={hideLabel}
        >
          {labelText}
        </FormLabel>
        }

        {controlsBlock}
      </div>
    );

    // Switch markup strategies for certain non-standard components
    if (choiceGroup) {
      groupBlock = (
        <Fieldset 
          className={variantClasses}
          legendText={labelText}
          legendClassName="control__label"
        >
          {controlsBlock}
        </Fieldset>
      );
    }
    else if (toggle) {
      groupBlock = (
        <div className={variantClasses}>
          <FormLabel 
            className="toggle-switch-wrapper-label" 
            required={required} 
            idGuid={idGuid}
          >
            <span className="control__label">
              {labelText}
            </span>
            {controlsBlock}
          </FormLabel>
        </div>
      );
    } 
    else if (inlineCheckbox || inlineRadioButton) {
      groupBlock = (
        <div className={variantClasses}>
          {controlsBlock}
          {labelText}
        </div>
      );
    }

    return (
      <React.Fragment>{groupBlock}</React.Fragment>
    );
  }
}