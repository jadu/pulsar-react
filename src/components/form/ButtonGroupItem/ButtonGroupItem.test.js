'use strict';

import React from 'react';
import { shallow, render } from 'enzyme';


import FormGroup from '../FormGroup';
import ButtonGroupItem from './ButtonGroupItem';

it('renders the basic input, defaulting to radio', () => {
  const tree = render(
    <ButtonGroupItem />
  );
  expect(tree).toMatchSnapshot();
});

it('renders the basic input, as checkbox', () => {
  const tree = render(
    <ButtonGroupItem type="checkbox" />
  );
  expect(tree).toMatchSnapshot();
});

it('renders the basic input with labelText', () => {
  const tree = render(
    <ButtonGroupItem>My Label</ButtonGroupItem>
  );
  expect(tree).toMatchSnapshot();
});

it('renders the basic input with name', () => {
  const tree = render(
    <ButtonGroupItem name="foo">My Label</ButtonGroupItem>
  );
  expect(tree).toMatchSnapshot();
});

it('renders the basic input with id', () => {
  const tree = render(
    <ButtonGroupItem id="foo">My Label</ButtonGroupItem>
  );
  expect(tree).toMatchSnapshot();
});

it('renders the basic input with idGuid', () => {
  const tree = render(
    <ButtonGroupItem idGuid="bar">My Label</ButtonGroupItem>
  );
  expect(tree).toMatchSnapshot();
});

it('will prefer id over idGuid', () => {
  const tree = render(
    <ButtonGroupItem id="foo" idGuid="bar">My Label</ButtonGroupItem>
  );
  expect(tree).toMatchSnapshot();
});

it('renders the basic input with aria-describedby', () => {
  const tree = render(
    <ButtonGroupItem aria-describedby="foo">My Label</ButtonGroupItem>
  );
  expect(tree).toMatchSnapshot();
});

it('renders as form group', () => {
  const tree = shallow(
    <FormGroup buttonGroup labelText="My Label">
      <>
        <ButtonGroupItem id="foo-1" name="group-1">Foo</ButtonGroupItem>
        <ButtonGroupItem id="bar-1" name="group-1">Bar</ButtonGroupItem>
      </>
    </FormGroup>
  );
  expect(tree).toMatchSnapshot();
});

it('renders as form group, with helpText, using the aria-describedby attribute', () => {
  const tree = shallow(
    <FormGroup buttonGroup labelText="My Label" helpText="foo">
      <>
        <ButtonGroupItem name="group-1">Foo</ButtonGroupItem>
      </>
    </FormGroup>
  );
  expect(tree).toMatchSnapshot();
});

it('renders as form group, with error, using the aria-describedby attribute', () => {
  const tree = shallow(
    <FormGroup buttonGroup labelText="My Label" error="bar">
      <>
        <ButtonGroupItem id="foo-1" name="group-1">Foo</ButtonGroupItem>
        <ButtonGroupItem id="bar-1" name="group-1">Bar</ButtonGroupItem>
      </>
    </FormGroup>
  );
  expect(tree).toMatchSnapshot();
});

it('renders as form group, with helpText and error, using two aria-describedby attributes', () => {
  const tree = shallow(
    <FormGroup buttonGroup labelText="My Label" helpText="foo" error="bar">
      <>
        <ButtonGroupItem id="foo-1" name="group-1">Foo</ButtonGroupItem>
        <ButtonGroupItem id="bar-1" name="group-1">Bar</ButtonGroupItem>
      </>
    </FormGroup>
  );
  expect(tree).toMatchSnapshot();
});