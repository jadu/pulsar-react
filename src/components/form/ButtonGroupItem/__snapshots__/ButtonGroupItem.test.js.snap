// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders as form group 1`] = `
<Fragment>
  <div
    className="form__group form__button-group"
  >
    <FormLabel
      className="control__label"
      idGuid="id-guid-1"
      required={false}
      tag="label"
    >
      My Label
    </FormLabel>
    <div
      className="controls btn__group"
    >
      <ButtonGroupItem
        ariaDescribedby={null}
        className="form__control"
        id="foo-1"
        idGuid="id-guid-1"
        key=".0"
        name="group-1"
        required={false}
        type="radio"
      >
        Foo
      </ButtonGroupItem>
      <ButtonGroupItem
        ariaDescribedby={null}
        className="form__control"
        id="bar-1"
        idGuid="id-guid-1"
        key=".1"
        name="group-1"
        required={false}
        type="radio"
      >
        Bar
      </ButtonGroupItem>
      <ErrorBlock
        errorGuid="error-guid-1"
      />
      <HelpBlock
        helpGuid="help-guid-1"
      />
    </div>
  </div>
</Fragment>
`;

exports[`renders as form group, with error, using the aria-describedby attribute 1`] = `
<Fragment>
  <div
    className="form__group form__button-group has-error"
  >
    <FormLabel
      className="control__label"
      idGuid="id-guid-1"
      required={false}
      tag="label"
    >
      My Label
    </FormLabel>
    <div
      className="controls btn__group"
    >
      <ButtonGroupItem
        ariaDescribedby="error-guid-1"
        className="form__control"
        id="foo-1"
        idGuid="id-guid-1"
        key=".0"
        name="group-1"
        required={false}
        type="radio"
      >
        Foo
      </ButtonGroupItem>
      <ButtonGroupItem
        ariaDescribedby="error-guid-1"
        className="form__control"
        id="bar-1"
        idGuid="id-guid-1"
        key=".1"
        name="group-1"
        required={false}
        type="radio"
      >
        Bar
      </ButtonGroupItem>
      <ErrorBlock
        errorGuid="error-guid-1"
      >
        bar
      </ErrorBlock>
      <HelpBlock
        helpGuid="help-guid-1"
      />
    </div>
  </div>
</Fragment>
`;

exports[`renders as form group, with helpText and error, using two aria-describedby attributes 1`] = `
<Fragment>
  <div
    className="form__group form__button-group has-error"
  >
    <FormLabel
      className="control__label"
      idGuid="id-guid-1"
      required={false}
      tag="label"
    >
      My Label
    </FormLabel>
    <div
      className="controls btn__group"
    >
      <ButtonGroupItem
        ariaDescribedby="error-guid-1 help-guid-1"
        className="form__control"
        id="foo-1"
        idGuid="id-guid-1"
        key=".0"
        name="group-1"
        required={false}
        type="radio"
      >
        Foo
      </ButtonGroupItem>
      <ButtonGroupItem
        ariaDescribedby="error-guid-1 help-guid-1"
        className="form__control"
        id="bar-1"
        idGuid="id-guid-1"
        key=".1"
        name="group-1"
        required={false}
        type="radio"
      >
        Bar
      </ButtonGroupItem>
      <ErrorBlock
        errorGuid="error-guid-1"
      >
        bar
      </ErrorBlock>
      <HelpBlock
        helpGuid="help-guid-1"
      >
        foo
      </HelpBlock>
    </div>
  </div>
</Fragment>
`;

exports[`renders as form group, with helpText, using the aria-describedby attribute 1`] = `
<Fragment>
  <div
    className="form__group form__button-group"
  >
    <FormLabel
      className="control__label"
      idGuid="id-guid-1"
      required={false}
      tag="label"
    >
      My Label
    </FormLabel>
    <div
      className="controls btn__group"
    >
      <ButtonGroupItem
        ariaDescribedby="help-guid-1"
        className="form__control"
        idGuid="id-guid-1"
        key=".0"
        name="group-1"
        required={false}
        type="radio"
      >
        Foo
      </ButtonGroupItem>
      <ErrorBlock
        errorGuid="error-guid-1"
      />
      <HelpBlock
        helpGuid="help-guid-1"
      >
        foo
      </HelpBlock>
    </div>
  </div>
</Fragment>
`;

exports[`renders the basic input with aria-describedby 1`] = `
Array [
  <input
    aria-describedby="foo"
    class="form__control radio"
    type="radio"
  />,
  <label
    class="control__label"
  >
    My Label
  </label>,
]
`;

exports[`renders the basic input with id 1`] = `
Array [
  <input
    class="form__control radio"
    id="foo"
    type="radio"
  />,
  <label
    class="control__label"
    for="foo"
  >
    My Label
  </label>,
]
`;

exports[`renders the basic input with idGuid 1`] = `
Array [
  <input
    class="form__control radio"
    id="bar"
    type="radio"
  />,
  <label
    class="control__label"
    for="bar"
  >
    My Label
  </label>,
]
`;

exports[`renders the basic input with labelText 1`] = `
Array [
  <input
    class="form__control radio"
    type="radio"
  />,
  <label
    class="control__label"
  >
    My Label
  </label>,
]
`;

exports[`renders the basic input with name 1`] = `
Array [
  <input
    class="form__control radio"
    name="foo"
    type="radio"
  />,
  <label
    class="control__label"
  >
    My Label
  </label>,
]
`;

exports[`renders the basic input, as checkbox 1`] = `
Array [
  <input
    class="form__control checkbox"
    type="checkbox"
  />,
  <span
    class="control__label"
  />,
]
`;

exports[`renders the basic input, defaulting to radio 1`] = `
Array [
  <input
    class="form__control radio"
    type="radio"
  />,
  <span
    class="control__label"
  />,
]
`;

exports[`will prefer id over idGuid 1`] = `
Array [
  <input
    class="form__control radio"
    id="foo"
    type="radio"
  />,
  <label
    class="control__label"
    for="foo"
  >
    My Label
  </label>,
]
`;
